firmware/raspi/boardinterface.py

```diff
@@ -170,44 +207,93 @@ def set_status_from_arduino(self, arduino_status=ArduinoStatus.IDLE):
+        # 8 x 8 chess board
+        for i in range(8):
+            # Print row, then number indicating rank
+            for j in range(8):
+                print(chess_grid[i][j], end=' ')
+            print(8-i)
+
+        # Print letters under board to indicate files
         for char in ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h']:
             print(char, end='')
-            print("   ", end='')
+            print(" ", end='')
         print()
 
     # def setup_board(self, is_human_turn):
     #     ''' Set up board with black or white on human side, depending on value of is_human_turn.
 
     #     If is_human_turn == True, white on human side; otherwise, black
     #     '''
-    #     # TODO: implement this method
     #     print("Need to discuss how board is setup in team meeting")
 
-    def handle_promotion(self, color, piece_type):
-        '''Backfills promotion area from graveyard if possible.
+    def handle_promotion(self, square, color, piece_type):
+        '''Assumes that handling captures during promotion done outside this function.
+
+        Backfills promotion area from graveyard if possible.
         '''
+        # TODO: First move the pawn being promoted to the graveyard.
+        # self.send_to_graveyard(color, 'p', origin=square)
+
+        # TODO: Put the to-be-promoted piece on the appropriate square.
+        # self.retrieve_from_graveyard(destination, color, piece_type)
+
+        # Lastly, backfill the graveyard, if possible.
         if self.graveyard.dead_piece_counts[color + piece_type] > 1:
             self.send_message_to_arduino(
                 *self.graveyard.backfill_promotion_area_from_graveyard(color, piece_type),
                 OpCode.MOVE_PIECE_ALONG_SQUARE_EDGES)
-            self.graveyard.update_dead_piece_count(color, piece_type, delta=1)
+            self.graveyard.update_dead_piece_count(color, piece_type, delta=-1)  # decrement
         else:
             print(f"All pieces of type {color}{piece_type} have been used!")
 
+    # TODO: clean up interaction between this and backfill_promo_area in Graveyard. Need to be
+    # careful about updating counts of dead pieces.
+
+    def send_to_graveyard(self, color, piece_type, origin=None):
+        '''Send piece to graveyard and increment dead piece count.
+
+        Increments the number of dead pieces from k to k + 1, then sends captured piece to
+        graveyard[piece types][k + 1].
+        '''
+        self.graveyard.update_dead_piece_count(color, piece_type, delta=1)  # increment
+        counts, graveyard = get_graveyard_info_for_piece_type(color, piece_type)
+        count = counts[color + piece_type]
+        if not origin:
+            origin = self.graveyard.w_capture_sq if color == 'w' else self.graveyard.b_capture_sq
+        dest = graveyard[color + piece_type][count]
+
+        self.send_message_to_arduino(origin, dest, OpCode.MOVE_PIECE_ALONG_SQUARE_EDGES)
+
+    def retrieve_from_graveyard(dest, color, piece_type):
```

Comment:
retrieve_from_graveyard function seems to increment the dead piece count, which I believe should decrease since you are retrieving a piece from the graveyard.

Reply:
Apart from that everything looks great.

Reply:
Thanks for the comments, you are absolutely right. Fixed.

Reply:
Sweet, looks good to pull.